### PR TITLE
Complete test coverage for `update-bcd` script

### PR DESCRIPTION
**Pull request by jugglinmike**
As we prepare to fix issues with the `update-bcd` script's "mirror" feature (see gh-2280), I felt it was important to have full test coverage for the script.

### CODE REVIEW

**Comment by jugglinmike** [comment]
I just pushed up a commit to also verify the return value of the `update` function.

**Comment by jugglinmike** [comment]
One of these tests was inaccurate, but it turns out that the corrected version exercises a bug and subsequently fails. Since this patch is intended to improve test coverage only, and since modifying the application logic will obscure that intent, I've opened gh-2326 to correct the bug. I'll be happy to rebase this patch once we merge that fix.

**Comment by foolip** [comment]
@jugglinmike I've merged https://github.com/foolip/mdn-bcd-collector/pull/2326 now, if you want to revisit this.

**Comment by jugglinmike** [comment]
@foolip Thanks for the review! I've resolved the conflicts with a merge commit. While this allowed me to avoid force-pushing and disrupting the review process, it also gummed up the commit history. The easiest resolution is to squash this all together while merging, but that might not be desirable.

Part of this work re-arranges tests. That kind of change is prone to silent coverage regression, so some maintainers would prefer to track it with a targeted commit.

If you'd prefer to structure the patch according to the commits I originally submitted, let me know, and I'll be happy to rebase and force-push.

File: unittest/unit/update-bcd.ts
Comment (foolip) [inline unittest/unit/update-bcd.ts:1278]: 👍

**Review by foolip — CHANGES_REQUESTED** [review]

**Comment by foolip** [comment]
Great tests, thanks @jugglinmike! I've asked to tweak some test names and to clone some tests to reflect my opinion about how the collector should change in the future. Feel free to only make the minimal edits to test names (no new tests) if you feel like it, up to you.

**Comment by foolip** [comment]
@queengooborg FYI, these tests (and my review comments) show why we can't really rely on `update-bcd` to confidently say that BCD is consistent with test results, given all the bugs we have.

File: unittest/unit/update-bcd.ts
Comment (jugglinmike) [reply unittest/unit/update-bcd.ts:794]: That's my career in a nutshell.

File: unittest/unit/update-bcd.ts
Comment (jugglinmike) [reply unittest/unit/update-bcd.ts:1039]: Better still: "retains flag data" (since "leaves" could be taken to mean "removes", as in "leaves behind").

File: unittest/unit/update-bcd.ts
Comment (jugglinmike) [reply unittest/unit/update-bcd.ts:1077]: This is a bit tricky: I think the situation you're describing is inaccurate, but your concern is valid and does indeed warrant a new test--just not the test you've suggested.

The test titled "skips removed features" exercises the branch you've highlighted. That test ought to fail when the corresponding TODO comment is addressed.

On the other hand, the test named "skips complex support scenarios" hits the branch immediately before the one you've highlighted:

https://github.com/foolip/mdn-bcd-collector/blob/19a623062ab0a82da8d3cc7c757452a5e408f877/update-bcd.ts#L381-L384

(The "TODO" comment there inspired the name of this test.) You're right that this will continue to pass after the "TODO" comment is addressed, so I agree that I should add another test which would fail when the skipping is replaced with more advanced logic. That test needs two "default statements" (to use the terminology of the source code), so I've selected a "support" value of: `[{version_added: '93'}, {version_removed: '94'}]`. When the "update" script is extended to handle this situation, it ought to change `version_added`, and the new test ought to fail.

File: unittest/unit/update-bcd.ts
Comment (jugglinmike) [reply unittest/unit/update-bcd.ts:1083]: Ah, I didn't realize that formulation was invalid. Your correction will degrade test coverage, though: these tests are specifically designed to exercise cases with two or more "default" support statements (where, in the parlance of the source code, "default" means "not bearing `flags`, `prefix`, or `alternative_name`). In order to make the data valid and preserve coverage, I've replaced the `version_removed` statement with one using `partial_implementation`.

File: unittest/unit/update-bcd.ts
Comment (jugglinmike) [reply unittest/unit/update-bcd.ts:1121]: Since I couldn't apply your suggestion to the above test, I can take that more direct remediation here.

File: unittest/unit/update-bcd.ts
Comment (jugglinmike) [reply unittest/unit/update-bcd.ts:1153]: :)

File: unittest/unit/update-bcd.ts
Comment (jugglinmike) [reply unittest/unit/update-bcd.ts:1201]: It looks like that test would fail in the current implementation. This block of code appears to defer to the existing information in BCD:

https://github.com/foolip/mdn-bcd-collector/blob/19a623062ab0a82da8d3cc7c757452a5e408f877/update-bcd.ts#L400-L417

So I think that this would be best addressed in a separate patch.

Comment (foolip) [reply unittest/unit/update-bcd.ts:1245]: Wow this code is very hard to even parse. It's kind of weird that we use `versionMap` here when we've already used it to create `inferredStatements`. But yeah it looks like we're just going to bail if we have any non-null result from any browser version older than what we have in BCD. That happened in https://github.com/foolip/mdn-bcd-collector/pull/1904.

I agree sorting that out in this PR doesn't make sense.

Comment (foolip) [reply unittest/unit/update-bcd.ts:1245]: I've filed https://github.com/foolip/mdn-bcd-collector/issues/2348.

**Review by foolip — APPROVED** [review]

**Pull request merged**
